plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -626,7 +667,22 @@ enterComputeCek costs = computeCek where
       case resultOrErr of
           -- Turn the cause of a possible failure, being a 'CekValue', into a 'Term'.
           Left e       -> throwCek $ mapCauseInMachineException (void . dischargeCekValue) e
-          Right result -> returnCek ctx result
+          Right result -> returnCek unbudgetedSteps ctx result
+
+    -- | Spend the budget that has been accumulated for a number of machine steps.
+    spendAccumulatedBudget :: Int -> CekM s ()
+    spendAccumulatedBudget !unbudgetedSteps = spendBudgetCek BStep (stimes unbudgetedSteps (cekStepCost ?cekCosts))
+
+    -- This will probably be inlined regardless, but it's important that it happen, since otherwise the return value
+    -- in CekM s Int won't be unboxed (although it seems like it could be).
+    {-# INLINE stepAndMaybeSpend #-}
+    -- | Accumulate a step, and maybe spend the budget that has accumulated for a number of machine steps, but only if we've exceeded our slippage.
+    stepAndMaybeSpend :: Int -> CekM s Int
+    stepAndMaybeSpend !unbudgetedSteps =
+        let !unbudgetedSteps' = unbudgetedSteps+1
+        in if unbudgetedSteps' >= ?cekSlippage
+        then spendAccumulatedBudget unbudgetedSteps' >> pure 0
+        else pure unbudgetedSteps'
```

Comment:
It's hard to preserve unboxing and be general over what kind of stuff we can do with costing, so maybe indeed add an `ExBudgetCategory` as an argument and ignore it by default, but use it when calibrating and control the mode via CPP?

Reply:
> maybe indeed add an `ExBudgetCategory` as an argument and ignore it by default, but use it when calibrating and control the mode via CPP?

... and then we'd need to recompile when we wanted to look more closely at a program?  I'm beginning to think that the easiest thing to do would be to have a separate evaluator for doing things like that.  It wouldn't have to be remotely efficient, so we could keep it very simple.  It wouldn't need all these safe/unsafe entry points for example.